design/002-secret-generator-group.md

```diff
@@ -0,0 +1,277 @@
+```yaml
+---
+title: Secret Generators
+version: v1alpha1
+authors: Christian Hünig, Jan Steffen, Moritz Johner
+creation-date: 2022-07-08
+status: draft
+---
+```
+
+# Secret Generators
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+## Summary
+
+There are use-cases where a secret could be generated by the operator instead of using pre-existing ones from the secret store. For example a MongoDB could be spun up using a password generated by the operator. This password is then used by an application in the cluster to access the database. No human ever needs to read that secret or have to specify it beforehand. A similar feature is available in a project called [VaultOperator](https://github.com/finleap-connect/vaultoperator/) which tries to solve similar problems like this operator just for Hashicorp Vault as a secret store.
+
+## Motivation
+
+The motivation is to boost adoption of zero-trust secrets management
+and to make it simple to create secure secrets for machine users
+of services running inside clusters (i.e. databases etc).
+
+### Goals
+
+* enable zero-trust secret management for purely technical and internal users
+* enable secure passwords by default
+* reduce error rates when creating RSA or ECDSA private keys
+
+### Non-Goals
+
+* auto-rotation of secrets
+
+## Proposal
+
+Implement several generators to support not only a number of data types but also generators that interface with external APIs.
+
+### API
+
+We add a new resource group `generator.external-secrets.io`. For each generator we add a new CRD, e.g. for password generation, VaultPKI, ECR, GCR etc.
+This custom resource contains all the configuration details to create a secure value.
+It can be referenced from a `Kind=ExternalSecret` or `Kind=PushSecret`. **or embedded** into it.
+
+The controller generates the value and passes it onto the `Kind=Secret` with all the values
+that have been fetched from the provider.
+When used with a `Kind=PushSecret` it will generate the value and push it to the configured provider.
+
+A generator generates a known set of outputs, e.g. `username` and `password`, `private_key` and `certificate` or others.
+
+#### Example CRD
+
+The following custom resource `Kind=VaultPKIBackend` contains all the parameters for issuing a certificate via HashiCorp Vault PKI engine.
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: VaultPKIBackend
+metadata:
+  name: my-vault-pki-creds
+spec:
+  # specific params for generating
+  commonName: example.com
+  altNames:
+  - example.org
+  - localhost
+  # ... more opts
+  auth: {} # service account etc. etc.
+```
+
+This resource can be **embedded or referenced** from an ExternalSecret or PushSecret.
+The ExternalSecret controller generates the secret value and it can be post-processed with
+existing features like template or key-rewrite.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
```

Comment:
`secretStoreRef` in `ExternalSecrets` and `SecretStoreRefs` in  `PushSecret` now feels obsolete to me! I think we can use the general idea and have a `sourceRef` per key in the `data/dataFrom` operations, where the possibilities can be `secretStoreRef/generator/generatorRef` for ExternalSecrets and `secretRef/generator/generatorRef` for secretStores.

Reply:
We can maintain the field and use it as the default if no `sourceRef` is provided. We can then deprecate it afterwards. WDYT?

Reply:
> [...] now feels obsolete to me!

Totally!

If i understand you right a `ExternalSecret` could then point to many `SecretStores`? :bird: :bird: :rock:
I like that idea! But would a user then have to add a `sourceRef` to each and every `key` :unamused: ? That would add quite a lot of overhead. I like the simplicity of having a 1:1 relationship of an ExternalSecret/SecretStore. 

Alternatively we can make the `SecretStore` optional :shrug:

Reply:
My original thought was to maintain the `SecretStoreRef` and use it as the default value if `sourceRef` is empty. This would allow the best of both worlds, plus it would allow us to make a seamless upgrade :smile:

Reply:
makes sense! While i like the idea, i fear that this is a little too magical for a first time user :thinking:. I guess as long as we have proper documentation it will be fine!

Reply:
While implementing it, sourceRef looks very weird on the generator spec.

This is how i would like it to look like:

```yaml
spec:
  dataFrom:

  # simple find
  - find:
      name:
        # match secret name with regexp
        regexp: "key-.*"

  # simple extract
  - extract:
      key: /dev/config
      property: foo.nested

  # simple generator   
  - generator:
      apiVersion: generators.external-secrets.io/v1alpha1
      kind: ACRAccessToken
      spec: {} # some inline spec

  # simple generatorRef 
  - generatorRef:
      apiVersion: generators.external-secrets.io/v1alpha1
      kind: ACRAccessToken
      name: "foo" 
```

Now, adding a `sourceRef` to the dataFrom[] type looks weird:
1. `sourceRef.secretStoreRef` applies to find/extract only
2. having `sourceRef.generator|Ref` can not / should not be used with find/extract
3. having `data[].sourceRef.generator|Ref` doesn't make sense because these would be two different instances of a generator.

```yaml
spec:
  dataFrom:
  # looks good with sourceRef
  - sourceRef:
      secretStoreRef: {...}
    find:
      name:
        # match secret name with regexp
        regexp: "key-.*"
  
  # simple generator   
  - generator:
      apiVersion: generators.external-secrets.io/v1alpha1
      kind: ACRAccessToken
      spec: {} # some inline spec

  # simple generatorRef 
  - generatorRef:
      apiVersion: generators.external-secrets.io/v1alpha1
      kind: ACRAccessToken
      name: "foo" 
```

With that being said i feel that we should drop `sourceRef` from this proposal and just make `spec.secretStoreRef` optional and add validation logic accordingly. WDYT?